File: flutter/lib/core/model/filters.dart
import 'dart:math' as math;

/// Looks are pure RGB to RGB functions, baked into a 64³ LUT at runtime.
///
/// Nothing binary ships with the app, so every look stays readable and
/// editable in source and costs nothing in APK size. It also means one
/// definition serves both consumers: the preview shader samples the baked LUT,
/// and export writes the same LUT out as a .cube for FFmpeg's lut3d filter.
/// A look cannot render differently on screen and in the file, because there
/// is only one of it.
class FilterPreset {
  const FilterPreset(this.id, this.name, this.category, this.look);

  final String id;
  final String name;
  final FilterCategory category;

  /// Transforms `rgb` (each 0..1) in place.
  final void Function(List<double>) look;
}

enum FilterCategory {
  basic('Basic'),
  portrait('Portrait'),
  film('Film'),
  cinema('Cinema'),
  mono('B&W'),
  vibe('Vibe');

  const FilterCategory(this.label);

  final String label;
}

/// The colour operations the presets are composed from.
class Look {
  const Look._();

  static double clamp01(double v) => v < 0 ? 0 : (v > 1 ? 1 : v);

  static double luma(List<double> c) =>
      0.2126 * c[0] + 0.7152 * c[1] + 0.0722 * c[2];

  static void contrast(List<double> c, double amount) {
    for (var i = 0; i < 3; i++) {
      c[i] = clamp01((c[i] - 0.5) * (1 + amount) + 0.5);
    }
  }

  static void saturate(List<double> c, double amount) {
    final l = luma(c);
    for (var i = 0; i < 3; i++) {
      c[i] = clamp01(l + (c[i] - l) * (1 + amount));
    }
  }

  static void gamma(List<double> c, double g) {
    for (var i = 0; i < 3; i++) {
      c[i] = clamp01(math.pow(math.max(c[i], 0.0), g).toDouble());
    }
  }

  /// Tints shadows and highlights in opposite directions.
  static void splitTone(
    List<double> c,
    List<double> shadow,
    List<double> highlight,
    double strength,
  ) {
    final l = luma(c);
    final hi = smoothstep(0.45, 1, l);
    final lo = 1 - smoothstep(0, 0.55, l);
    for (var i = 0; i < 3; i++) {
      c[i] = clamp01(
        c[i] + (shadow[i] - 0.5) * lo * strength + (highlight[i] - 0.5) * hi * strength,
      );
    }
  }

  static void mono(
    List<double> c, {
    double rw = 0.299,
    double gw = 0.587,
    double bw = 0.114,
  }) {
    final v = clamp01(c[0] * rw + c[1] * gw + c[2] * bw);
    c[0] = v;
    c[1] = v;
    c[2] = v;
  }

  static void temperature(List<double> c, double amount) {
    c[0] = clamp01(c[0] + amount * 0.14);
    c[2] = clamp01(c[2] - amount * 0.14);
  }

  /// Lifts the black point, the defining move of a faded or matte look.
  static void fade(List<double> c, double amount) {
    for (var i = 0; i < 3; i++) {
      c[i] = clamp01(c[i] * (1 - amount * 0.18) + amount * 0.16);
    }
  }

  static double smoothstep(double e0, double e1, double x) {
    final t = ((x - e0) / (e1 - e0)).clamp(0.0, 1.0);
    return t * t * (3 - 2 * t);
  }

  /// Filmic S-curve; `amount` of 0 is identity.
  static void sCurve(List<double> c, double amount) {
    for (var i = 0; i < 3; i++) {
      final x = c[i];
      final s = x * x * (3 - 2 * x);
      c[i] = clamp01(x + (s - x) * amount);
    }
  }

  static void channelMix(List<double> c, double rr, double gg, double bb) {
    c[0] = clamp01(c[0] * rr);
    c[1] = clamp01(c[1] * gg);
    c[2] = clamp01(c[2] * bb);
  }
}

class Filters {
  const Filters._();

  static const String noneId = 'none';

  static final List<FilterPreset> all = [
    const FilterPreset(noneId, 'Original', FilterCategory.basic, _original),

    const FilterPreset('crisp', 'Crisp', FilterCategory.basic, _crisp),
    const FilterPreset('soft', 'Soft', FilterCategory.basic, _soft),
    const FilterPreset('punch', 'Punch', FilterCategory.basic, _punch),

    const FilterPreset('natural', 'Natural', FilterCategory.portrait, _natural),
    const FilterPreset('glow', 'Glow', FilterCategory.portrait, _glowLook),
    const FilterPreset('honey', 'Honey', FilterCategory.portrait, _honey),
    const FilterPreset(
        'porcelain', 'Porcelain', FilterCategory.portrait, _porcelain),

    const FilterPreset('kodachrome', 'Kodak', FilterCategory.film, _kodachrome),
    const FilterPreset('portra', 'Portra', FilterCategory.film, _portra),
    const FilterPreset('fuji', 'Fuji', FilterCategory.film, _fuji),
    const FilterPreset('expired', 'Expired', FilterCategory.film, _expired),

    const FilterPreset(
        'teal_orange', 'Blockbuster', FilterCategory.cinema, _tealOrange),
    const FilterPreset('noir_blue', 'Midnight', FilterCategory.cinema, _noirBlue),
    const FilterPreset('desert', 'Desert', FilterCategory.cinema, _desert),
    const FilterPreset('neon', 'Neon', FilterCategory.cinema, _neon),

    const FilterPreset('mono', 'Mono', FilterCategory.mono, _mono),
    const FilterPreset('mono_hard', 'High key', FilterCategory.mono, _monoHard),
    const FilterPreset('mono_soft', 'Silver', FilterCategory.mono, _monoSoft),
    const FilterPreset('sepia', 'Sepia', FilterCategory.mono, _sepia),

    const FilterPreset('vhs', 'VHS', FilterCategory.vibe, _vhs),
    const FilterPreset('dream', 'Dream', FilterCategory.vibe, _dream),
    const FilterPreset('cotton', 'Cotton', FilterCategory.vibe, _cotton),
    const FilterPreset('acid', 'Acid', FilterCategory.vibe, _acid),
  ];

  static final Map<String, FilterPreset> byId = {
    for (final preset in all) preset.id: preset,
  };

  static FilterPreset get(String id) => byId[id] ?? all.first;

  static List<FilterPreset> inCategory(FilterCategory category) =>
      all.where((preset) => preset.category == category).toList();
}

void _original(List<double> c) {}

void _crisp(List<double> c) {
  Look.contrast(c, 0.14);
  Look.saturate(c, 0.10);
  Look.sCurve(c, 0.25);
}

void _soft(List<double> c) {
  Look.contrast(c, -0.10);
  Look.fade(c, 0.35);
  Look.saturate(c, -0.05);
}

void _punch(List<double> c) {
  Look.contrast(c, 0.28);
  Look.saturate(c, 0.30);
  Look.sCurve(c, 0.35);
}

void _natural(List<double> c) {
  Look.temperature(c, 0.10);
  Look.saturate(c, 0.06);
  Look.sCurve(c, 0.15);
}

void _glowLook(List<double> c) {
  Look.temperature(c, 0.14);
  Look.fade(c, 0.20);
  Look.saturate(c, 0.08);
  Look.splitTone(c, const [0.5, 0.5, 0.54], const [0.56, 0.52, 0.46], 0.35);
}

void _honey(List<double> c) {
  Look.temperature(c, 0.22);
  Look.contrast(c, 0.08);
  Look.splitTone(c, const [0.48, 0.49, 0.55], const [0.60, 0.53, 0.42], 0.5);
}

void _porcelain(List<double> c) {
  Look.saturate(c, -0.18);
  Look.fade(c, 0.28);
  Look.contrast(c, 0.06);
  Look.temperature(c, 0.06);
}

void _kodachrome(List<double> c) {
  Look.contrast(c, 0.18);
  Look.saturate(c, 0.22);
  Look.channelMix(c, 1.04, 0.99, 0.92);
  Look.splitTone(c, const [0.47, 0.50, 0.56], const [0.57, 0.52, 0.44], 0.35);
}

void _portra(List<double> c) {
  Look.fade(c, 0.22);
  Look.saturate(c, -0.06);
  Look.temperature(c, 0.12);
  Look.splitTone(c, const [0.49, 0.50, 0.53], const [0.55, 0.51, 0.47], 0.4);
}

void _fuji(List<double> c) {
  Look.saturate(c, 0.14);
  Look.contrast(c, 0.12);
  Look.channelMix(c, 0.97, 1.03, 1.02);
}

void _expired(List<double> c) {
  Look.fade(c, 0.45);
  Look.saturate(c, -0.22);
  Look.splitTone(c, const [0.54, 0.50, 0.44], const [0.48, 0.52, 0.55], 0.55);
  Look.contrast(c, -0.06);
}

void _tealOrange(List<double> c) {
  Look.contrast(c, 0.16);
  Look.splitTone(c, const [0.42, 0.50, 0.60], const [0.60, 0.51, 0.40], 0.7);
  Look.saturate(c, 0.08);
}

void _noirBlue(List<double> c) {
  Look.contrast(c, 0.22);
  Look.saturate(c, -0.30);
  Look.splitTone(c, const [0.44, 0.48, 0.60], const [0.50, 0.51, 0.55], 0.6);
  Look.gamma(c, 1.12);
}

void _desert(List<double> c) {
  Look.temperature(c, 0.20);
  Look.saturate(c, -0.08);
  Look.contrast(c, 0.10);
  Look.splitTone(c, const [0.52, 0.50, 0.46], const [0.58, 0.53, 0.44], 0.45);
}

void _neon(List<double> c) {
  Look.contrast(c, 0.20);
  Look.saturate(c, 0.35);
  Look.splitTone(c, const [0.46, 0.46, 0.62], const [0.62, 0.46, 0.58], 0.6);
}

void _mono(List<double> c) {
  Look.mono(c);
  Look.contrast(c, 0.12);
}

void _monoHard(List<double> c) {
  Look.mono(c);
  Look.contrast(c, 0.42);
  Look.sCurve(c, 0.4);
}

void _monoSoft(List<double> c) {
  Look.mono(c);
  Look.fade(c, 0.35);
  Look.contrast(c, -0.05);
}

void _sepia(List<double> c) {
  Look.mono(c);
  c[0] = Look.clamp01(c[0] * 1.12 + 0.04);
  c[1] = Look.clamp01(c[1] * 1.00 + 0.01);
  c[2] = Look.clamp01(c[2] * 0.82);
}

void _vhs(List<double> c) {
  Look.fade(c, 0.4);
  Look.saturate(c, 0.18);
  Look.splitTone(c, const [0.54, 0.48, 0.56], const [0.48, 0.54, 0.52], 0.5);
}

void _dream(List<double> c) {
  Look.fade(c, 0.5);
  Look.saturate(c, -0.12);
  Look.temperature(c, 0.08);
  Look.gamma(c, 0.92);
}

void _cotton(List<double> c) {
  Look.fade(c, 0.32);
  Look.splitTone(c, const [0.55, 0.49, 0.53], const [0.53, 0.51, 0.56], 0.6);
  Look.saturate(c, 0.05);
}

void _acid(List<double> c) {
  Look.saturate(c, 0.55);
  Look.contrast(c, 0.24);
  Look.channelMix(c, 1.05, 1.02, 0.9);
}

File: android/app/src/main/java/com/vixel/studio/core/model/Filters.kt
package com.vixel.studio.core.model

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow

/**
 * Looks are defined as pure RGB -> RGB functions and baked into a 64^3 LUT at
 * runtime (see LutGenerator). No binary LUT assets ship with the app, so every
 * look is readable, tweakable, and costs nothing in APK size.
 */
data class FilterPreset(
    val id: String,
    val name: String,
    val category: Category,
    /** Transforms `rgb` (0..1) in place. */
    val look: (FloatArray) -> Unit,
) {
    enum class Category(val label: String) {
        BASIC("Basic"),
        PORTRAIT("Portrait"),
        FILM("Film"),
        CINEMA("Cinema"),
        MONO("B&W"),
        VIBE("Vibe"),
    }
}

/** Small library of colour operations the presets are composed from. */
object Look {
    fun clamp01(v: Float): Float = if (v < 0f) 0f else if (v > 1f) 1f else v

    fun luma(c: FloatArray): Float = 0.2126f * c[0] + 0.7152f * c[1] + 0.0722f * c[2]

    fun contrast(c: FloatArray, amount: Float) {
        for (i in 0..2) c[i] = clamp01((c[i] - 0.5f) * (1f + amount) + 0.5f)
    }

    fun saturate(c: FloatArray, amount: Float) {
        val l = luma(c)
        for (i in 0..2) c[i] = clamp01(l + (c[i] - l) * (1f + amount))
    }

    fun gamma(c: FloatArray, g: Float) {
        for (i in 0..2) c[i] = clamp01(c[i].coerceAtLeast(0f).pow(g))
    }

    /** Lift/gamma/gain per channel — the classic three-way colour balance. */
    fun lgg(c: FloatArray, lift: FloatArray, gamma: FloatArray, gain: FloatArray) {
        for (i in 0..2) {
            var v = c[i] * gain[i] + lift[i]
            v = clamp01(v).pow(gamma[i])
            c[i] = clamp01(v)
        }
    }

    /** Tints shadows and highlights in opposite directions. */
    fun splitTone(c: FloatArray, shadow: FloatArray, highlight: FloatArray, strength: Float) {
        val l = luma(c)
        val hi = smoothstep(0.45f, 1f, l)
        val lo = 1f - smoothstep(0f, 0.55f, l)
        for (i in 0..2) {
            c[i] = clamp01(c[i] + (shadow[i] - 0.5f) * lo * strength + (highlight[i] - 0.5f) * hi * strength)
        }
    }

    fun mono(c: FloatArray, rw: Float = 0.299f, gw: Float = 0.587f, bw: Float = 0.114f) {
        val v = clamp01(c[0] * rw + c[1] * gw + c[2] * bw)
        c[0] = v; c[1] = v; c[2] = v
    }

    fun temperature(c: FloatArray, amount: Float) {
        c[0] = clamp01(c[0] + amount * 0.14f)
        c[2] = clamp01(c[2] - amount * 0.14f)
    }

    /** Lifts the black point, the defining move of a faded/matte look. */
    fun fade(c: FloatArray, amount: Float) {
        for (i in 0..2) c[i] = clamp01(c[i] * (1f - amount * 0.18f) + amount * 0.16f)
    }

    fun smoothstep(e0: Float, e1: Float, x: Float): Float {
        val t = ((x - e0) / (e1 - e0)).coerceIn(0f, 1f)
        return t * t * (3f - 2f * t)
    }

    /** Filmic S-curve; `amount` 0 is identity. */
    fun sCurve(c: FloatArray, amount: Float) {
        for (i in 0..2) {
            val x = c[i]
            val s = x * x * (3f - 2f * x)
            c[i] = clamp01(x + (s - x) * amount)
        }
    }

    fun channelMix(c: FloatArray, rr: Float, gg: Float, bb: Float) {
        c[0] = clamp01(c[0] * rr)
        c[1] = clamp01(c[1] * gg)
        c[2] = clamp01(c[2] * bb)
    }
}

object Filters {
    const val NONE_ID = "none"

    private fun f(id: String, name: String, cat: FilterPreset.Category, look: (FloatArray) -> Unit) =
        FilterPreset(id, name, cat, look)

    val ALL: List<FilterPreset> = listOf(
        f(NONE_ID, "Original", FilterPreset.Category.BASIC) { },

        f("crisp", "Crisp", FilterPreset.Category.BASIC) { c ->
            Look.contrast(c, 0.14f); Look.saturate(c, 0.10f); Look.sCurve(c, 0.25f)
        },
        f("soft", "Soft", FilterPreset.Category.BASIC) { c ->
            Look.contrast(c, -0.10f); Look.fade(c, 0.35f); Look.saturate(c, -0.05f)
        },
        f("punch", "Punch", FilterPreset.Category.BASIC) { c ->
            Look.contrast(c, 0.28f); Look.saturate(c, 0.30f); Look.sCurve(c, 0.35f)
        },

        f("natural", "Natural", FilterPreset.Category.PORTRAIT) { c ->
            Look.temperature(c, 0.10f); Look.saturate(c, 0.06f); Look.sCurve(c, 0.15f)
        },
        f("glow", "Glow", FilterPreset.Category.PORTRAIT) { c ->
            Look.temperature(c, 0.14f); Look.fade(c, 0.20f); Look.saturate(c, 0.08f)
            Look.splitTone(c, floatArrayOf(0.5f, 0.5f, 0.54f), floatArrayOf(0.56f, 0.52f, 0.46f), 0.35f)
        },
        f("honey", "Honey", FilterPreset.Category.PORTRAIT) { c ->
            Look.temperature(c, 0.22f); Look.contrast(c, 0.08f)
            Look.splitTone(c, floatArrayOf(0.48f, 0.49f, 0.55f), floatArrayOf(0.60f, 0.53f, 0.42f), 0.5f)
        },
        f("porcelain", "Porcelain", FilterPreset.Category.PORTRAIT) { c ->
            Look.saturate(c, -0.18f); Look.fade(c, 0.28f); Look.contrast(c, 0.06f)
            Look.temperature(c, 0.06f)
        },

        f("kodachrome", "Kodak", FilterPreset.Category.FILM) { c ->
            Look.contrast(c, 0.18f); Look.saturate(c, 0.22f)
            Look.channelMix(c, 1.04f, 0.99f, 0.92f)
            Look.splitTone(c, floatArrayOf(0.47f, 0.50f, 0.56f), floatArrayOf(0.57f, 0.52f, 0.44f), 0.35f)
        },
        f("portra", "Portra", FilterPreset.Category.FILM) { c ->
            Look.fade(c, 0.22f); Look.saturate(c, -0.06f); Look.temperature(c, 0.12f)
            Look.splitTone(c, floatArrayOf(0.49f, 0.50f, 0.53f), floatArrayOf(0.55f, 0.51f, 0.47f), 0.4f)
        },
        f("fuji", "Fuji", FilterPreset.Category.FILM) { c ->
            Look.saturate(c, 0.14f); Look.contrast(c, 0.12f)
            Look.channelMix(c, 0.97f, 1.03f, 1.02f)
        },
        f("expired", "Expired", FilterPreset.Category.FILM) { c ->
            Look.fade(c, 0.45f); Look.saturate(c, -0.22f)
            Look.splitTone(c, floatArrayOf(0.54f, 0.50f, 0.44f), floatArrayOf(0.48f, 0.52f, 0.55f), 0.55f)
            Look.contrast(c, -0.06f)
        },

        f("teal_orange", "Blockbuster", FilterPreset.Category.CINEMA) { c ->
            Look.contrast(c, 0.16f)
            Look.splitTone(c, floatArrayOf(0.42f, 0.50f, 0.60f), floatArrayOf(0.60f, 0.51f, 0.40f), 0.7f)
            Look.saturate(c, 0.08f)
        },
        f("noir_blue", "Midnight", FilterPreset.Category.CINEMA) { c ->
            Look.contrast(c, 0.22f); Look.saturate(c, -0.30f)
            Look.splitTone(c, floatArrayOf(0.44f, 0.48f, 0.60f), floatArrayOf(0.50f, 0.51f, 0.55f), 0.6f)
            Look.gamma(c, 1.12f)
        },
        f("desert", "Desert", FilterPreset.Category.CINEMA) { c ->
            Look.temperature(c, 0.20f); Look.saturate(c, -0.08f); Look.contrast(c, 0.10f)
            Look.splitTone(c, floatArrayOf(0.52f, 0.50f, 0.46f), floatArrayOf(0.58f, 0.53f, 0.44f), 0.45f)
        },
        f("neon", "Neon", FilterPreset.Category.CINEMA) { c ->
            Look.contrast(c, 0.20f); Look.saturate(c, 0.35f)
            Look.splitTone(c, floatArrayOf(0.46f, 0.46f, 0.62f), floatArrayOf(0.62f, 0.46f, 0.58f), 0.6f)
        },

        f("mono", "Mono", FilterPreset.Category.MONO) { c ->
            Look.mono(c); Look.contrast(c, 0.12f)
        },
        f("mono_hard", "High key", FilterPreset.Category.MONO) { c ->
            Look.mono(c); Look.contrast(c, 0.42f); Look.sCurve(c, 0.4f)
        },
        f("mono_soft", "Silver", FilterPreset.Category.MONO) { c ->
            Look.mono(c); Look.fade(c, 0.35f); Look.contrast(c, -0.05f)
        },
        f("sepia", "Sepia", FilterPreset.Category.MONO) { c ->
            Look.mono(c)
            c[0] = Look.clamp01(c[0] * 1.12f + 0.04f)
            c[1] = Look.clamp01(c[1] * 1.00f + 0.01f)
            c[2] = Look.clamp01(c[2] * 0.82f)
        },

        f("vhs", "VHS", FilterPreset.Category.VIBE) { c ->
            Look.fade(c, 0.4f); Look.saturate(c, 0.18f)
            Look.splitTone(c, floatArrayOf(0.54f, 0.48f, 0.56f), floatArrayOf(0.48f, 0.54f, 0.52f), 0.5f)
        },
        f("dream", "Dream", FilterPreset.Category.VIBE) { c ->
            Look.fade(c, 0.5f); Look.saturate(c, -0.12f); Look.temperature(c, 0.08f)
            Look.gamma(c, 0.92f)
        },
        f("cotton", "Cotton", FilterPreset.Category.VIBE) { c ->
            Look.fade(c, 0.32f)
            Look.splitTone(c, floatArrayOf(0.55f, 0.49f, 0.53f), floatArrayOf(0.53f, 0.51f, 0.56f), 0.6f)
            Look.saturate(c, 0.05f)
        },
        f("acid", "Acid", FilterPreset.Category.VIBE) { c ->
            Look.saturate(c, 0.55f); Look.contrast(c, 0.24f)
            Look.channelMix(c, 1.05f, 1.02f, 0.9f)
        },
    )

    val BY_ID: Map<String, FilterPreset> = ALL.associateBy { it.id }

    fun get(id: String): FilterPreset = BY_ID[id] ?: ALL.first()

    fun categories(): List<FilterPreset.Category> = FilterPreset.Category.entries.toList()
}
